packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart

```diff
@@ -0,0 +1,342 @@
+  Future<Null> stop() async {
+    for (PortForwarder fp in _forwardedVmServicePorts) {
+      // Closes VM service first to ensure that the connection is closed cleanly
+      // on the target before shutting down the forwarding itself.
+      final DartVm vmService = _dartVmCache[fp.port];
+      _dartVmCache[fp.port] = null;
+      await vmService?.stop();
+      await fp.stop();
+    }
+    _dartVmCache.clear();
+    _forwardedVmServicePorts.clear();
+  }
+
+  /// Returns a list of [FlutterView] objects.
+  ///
+  /// This is run across all connected DartVM connections that this class is
+  /// managing.
+  Future<List<FlutterView>> getFlutterViews() async {
+    final List<FlutterView> views = <FlutterView>[];
+    if (_forwardedVmServicePorts.isEmpty) {
+      return views;
+    }
+    for (PortForwarder fp in _forwardedVmServicePorts) {
+      final DartVm vmService = await _getDartVm(fp.port);
+      views.addAll(await vmService.getAllFlutterViews());
+    }
+    return new List<FlutterView>.unmodifiable(views);
+  }
+
+  Future<DartVm> _getDartVm(int port) async {
+    if (!_dartVmCache.containsKey(port)) {
+      // While the IPv4 loopback can be used for the initial port forwarding
+      // (see [PortForwarder.start]), the address is actually bound to the IPv6
+      // loopback device, so connecting to the IPv4 loopback would fail when the
+      // target address is IPv6 link-local.
+      final String addr = _useIpV6Loopback
+          ? 'http://\[$_ipv6Loopback\]:$port'
+          : 'http://$_ipv4Loopback:$port';
+      final Uri uri = Uri.parse(addr);
+      final DartVm dartVm = await DartVm.connect(uri);
+      _dartVmCache[port] = dartVm;
+    }
+    return _dartVmCache[port];
+  }
+
+  /// Forwards a series of local device ports to the remote device.
+  ///
+  /// When this function is run, all existing forwarded ports and connections
+  /// are reset by way of [stop].
+  Future<Null> _forwardLocalPortsToDeviceServicePorts() async {
+    await stop();
+    final List<int> servicePorts = await getDeviceServicePorts();
+    _forwardedVmServicePorts
+        .addAll(await Future.wait(servicePorts.map((int deviceServicePort) {
+      return fuchsiaPortForwardingFunction(
+          _sshCommandRunner.address,
+          deviceServicePort,
+          _sshCommandRunner.interface,
+          _sshCommandRunner.sshConfigPath);
+    })));
+  }
+
+  /// Gets the open Dart VM service ports on a remote Fuchsia device.
+  ///
+  /// The method attempts to get service ports through an SSH connection. Upon
+  /// successfully getting the VM service ports, returns them as a list of
+  /// integers. If an empty list is returned, then no Dart VM instances could be
+  /// found. An exception is thrown in the event of an actual error when
+  /// attempting to acquire the ports.
+  Future<List<int>> getDeviceServicePorts() async {
+    final List<String> lsOutput =
+        await _sshCommandRunner.run('ls /tmp/dart.services');
```

Comment:
my understanding is that this feature's days are numbered. maybe add a TODO saying we'll soon have to fix this.

Reply:
Correct.  The plan was to figure out/change this shortly after submitting this change.  TODO added.